packages/eslint-plugin/tests/rules/no-confusing-void-expression.test.ts

```diff
@@ -119,6 +119,325 @@ function cool(input: string) {
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type Foo = {
+  foo: () => void;
+};
+const q = {
+  foo: () => console.log(),
+} as Foo;
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type Foo = {
+  foo: () => void;
+};
+const x: Foo = {
+  foo: () => console.log(),
+};
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+const q = {
+  foo: () => console.log(),
+} as {
+  foo: () => void;
+};
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+const x: {
+  foo: () => void;
+} = {
+  foo: () => console.log(),
+};
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type Foo = {
+  foo: { bar: () => void };
+};
+const q = {
+  foo: { bar: () => console.log() },
+} as Foo;
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type Foo = {
+  foo: { bar: () => void };
+};
+const x: Foo = {
+  foo: { bar: () => console.log() },
+};
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type MethodType = () => void;
+
+class App {
+  private method: MethodType = () => console.log();
+}
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+interface Foo {
+  foo: () => void;
+}
+function bar(): Foo {
+  return {
+    foo: () => console.log(),
+  };
+}
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type HigherOrderType = () => () => () => void;
+const x: HigherOrderType = () => () => () => console.log();
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+declare function foo(arg: () => void): void;
+foo(() => console.log());
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: 'const foo = <button onClick={() => console.log()} />;',
+      options: [{ ignoreVoidInVoid: true }],
+      parserOptions: {
+        ecmaFeatures: {
+          jsx: true,
+        },
+      },
+    },
+    {
+      code: `
+type HigherOrderType = () => () => () => void;
+const x: HigherOrderType = () => () => () => {
+  return console.log();
+};
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+type HigherOrderType = () => () => () => void;
+const x: HigherOrderType = () => () =>
+  function () {
+    return console.log();
+  };
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+function test1(): void;
+function test1(arg: string): any;
+function test1(arg?: string): any | void {
+  if (arg) {
+    return arg;
+  }
+
+  return console.log(); // not reported - ok
+}
+      `,
+      options: [{ ignoreVoidInVoid: true }],
+    },
+    {
+      code: `
+function test2(arg: string): any;
+function test2(): void;
+function test2(arg?: string): any | void {
+  if (arg) {
+    return arg;
+  }
+
+  return console.log(); // reported - bug
```

Comment:
Please see my comment from the one of the previous reviews https://github.com/typescript-eslint/typescript-eslint/pull/8632#discussion_r1539679818. :slightly_smiling_face: 

The code samples I provide, which contain some error messages, are for reference only. They are not intended to be directly copied into real tests. I'm including comments on these code samples to better illustrate what the bug I found is. The same goes for variable names: `test1` and `test2` are named to avoid name collisions in the playground. Having the `test2` function in a separate test case without `test1` looks a bit confusing.

I ask you to pay attention to the reviews and not just blindly insert some stuff into the code, assuming that someone will notice it in one of the next reviews anyway.

Reply:
I apologize for making the same mistake by not checking the code thoroughly after writing it.

In the future, I will look more closely before submitting a review.